### PR TITLE
Dynamic Binding Generation for Class Template

### PR DESCRIPTION
**Pull request by isVoid**
Dynamic binding generator for c++ class template. Demo:

Source C++ declarations:
```c++
template <int N, typename T> class Foo {
public:
  T t;
  __device__ Foo(T t) : t(t), t2(static_cast<T>(N)) {}

  __device__ T get_t() { return t; }
  __device__ T get_t2() { return t2; }

private:
  T t2;
};
```

Binding Usage Demo:
```python
@cuda.jit(link=shim_writer.links())
def kernel(out):
    # Stage 1: Instantiation. An API that provides an explicit instantiation
    # is provided.
    foo_t = Foo(T=np.int32, N=128)

    # Stage 2: Invoke constructor for instantiated type.
    foo = foo_t(256)

    # Stage 3: Use the constructed type.
    out[0] = foo.t
    out[1] = foo.get_t()
    out[2] = foo.get_t2()

# out == [256, 256, 128]
```

We apparently do not support every template syntax in c++, currently a subset of class template parameters are supported (type and non-type).

Additionally, we haven't provided support for template methods within a template class, or a nested templated struct.

### CODE REVIEW

File: ast_canopy/cpp/src/class_template_specialization.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    : Record(llvm::cast<clang::CXXRecordDecl>(CTSD),
+             RecordAncestor::ANCESTOR_IS_NOT_TEMPLATE),
+      class_template(CTSD->getSpecializedTemplate()) {
+
+  // Get the actual template arguments
+  auto &tparam_list = CTSD->getTemplateArgs();
+  actual_template_arguments.reserve(tparam_list.size());
+
+  for (auto i = 0; i < tparam_list.size(); i++) {
+    auto targ = tparam_list[i];
```
Comment (rparolin) [inline ast_canopy/cpp/src/class_template_specialization.cpp]: Consider: `const auto&` if targ is expensive to copy.

File: ast_canopy/cpp/src/class_template_specialization.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    clang::TemplateArgument::ArgKind kind = targ.getKind();
+    switch (kind) {
+    case clang::TemplateArgument::ArgKind::Type:
+      actual_template_arguments.push_back(targ.getAsType().getAsString());
+      break;
+    case clang::TemplateArgument::ArgKind::Integral: {
+      llvm::APSInt integer = targ.getAsIntegral();
+      llvm::SmallString<32> str; // -uint64_t::max() is 21 digits (with sign).
```
Comment (rparolin) [inline ast_canopy/cpp/src/class_template_specialization.cpp]: Unsigned 128-bit integer is 39 digits long. So you might want to bump this to 64-bytes and should be sufficient.

Comment (isVoid) [reply ast_canopy/cpp/src/class_template_specialization.cpp]: Does CUDA support 128-bit integers?

File: ast_canopy/tests/test_class_template_specialization.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    assert len(ctsd[0].templated_methods) == 0
+
+    # FIXME: Should be 0, CTSD creates an implicit Record for the underlying struct.
+    # It's actually the same record as the one specialized.
+    assert len(ctsd[0].nested_records) == 1
+
+    assert len(ctsd[0].nested_class_templates) == 0
+
+    # C++ standards says that an empty class has non-zero size.
+    assert ctsd[0].sizeof_ != 0
```
[the remark this answers is not in the document]
Comment (isVoid) [reply ast_canopy/tests/test_class_template_specialization.py]: Ah, good to know about this. Do you expect a test for this corner case to be created?

File: ast_canopy/cpp/src/class_template_specialization.cpp (hunk 1 above)
Comment (isVoid) [reply ast_canopy/cpp/src/class_template_specialization.cpp]: [b2269ef](https://github.com/NVIDIA/numbast/pull/244/commits/b2269ef8b0ac277a5402e3190d693b055cf3df16)

**Review by rparolin — APPROVED** [review]

File: numbast/src/numbast/class_template.py [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+        params=method_decl.params,
+    )
+
+    qualname = f"{s_type}.{method_decl.name}"
+
+    @lower(qualname, s_type, *param_types)
+    def _method_impl(context, builder, sig, args):
+        shim_writer.write_to_shim(shim, func_name)
+
+        # The first argument in argptrs is self, no need to extra allocate.
```
Comment (tpn) [inline numbast/src/numbast/class_template.py]: This comment is a bit confusing--it implies you're either not allocating for sig.args[0], or you're doing extra allocation for sig.args[1:], but... I don't see anything like that in the following code?  sig.args[0] appears to be treated the same as everything else?

File: numbast/src/numbast/class_template.py [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+                members = [
+                    (
+                        f.name,
+                        to_numba_type(f.type_.unqualified_non_ref_type_name),
+                    )
+                    for f in struct_decl.fields
+                ]
+                super().__init__(dmm, fe_type, members)
+
+    if data_model == StructModel:
```
Comment (tpn) [inline numbast/src/numbast/class_template.py]: Hmmm, why this extra block?  Why not combine with the elif block above?  Same predicate right?

File: numbast/src/numbast/class_template.py (hunk 4 above)
Comment (isVoid) [reply numbast/src/numbast/class_template.py]: Yes this is indeed a bit confusing. This is in contrast to the ctor lowering where the instance requires an additional allocation on stack. Though this should be self-evident because struct method does not involve in allocation.
